source/common/upstream/ring_hash_lb.cc

```diff
@@ -60,34 +66,97 @@ HostConstSharedPtr RingHashLoadBalancer::Ring::chooseHost(uint64_t h) const {
+  // weight, if given. Sum these effective weights.
+  uint64_t weighted_sum = 0;
+  std::unordered_map<HostConstSharedPtr, uint32_t> effective_weights;
+  for (uint64_t i = 0; i < hosts_per_locality.get().size(); ++i) {
+    for (const auto& host : hosts_per_locality.get()[i]) {
+      auto host_weight = host->weight();
+      ASSERT(host_weight != 0);
+      // NOTE: Locality weight may be explicitly set to zero, meaning any hosts in the locality
+      //       should be assigned no load.
+      // TODO: When we move to C++17, change this to `locality_weights[i]` (i.e. use
+      //       std::shared_ptr::operator[]) rather than dereferencing locality_weights explicitly.
+      auto locality_weight = locality_weights == nullptr ? 1 : (*locality_weights)[i];
+      if (locality_weight != 0) {
+        const auto effective_weight = host_weight * locality_weight;
+        weighted_sum += effective_weight;
+        effective_weights[host] = effective_weight;
+      }
     }
   }
 
-  ENVOY_LOG(info, "ring hash: min_ring_size={} hashes_per_host={}", min_ring_size, hashes_per_host);
-  ring_.reserve(hosts.size() * hashes_per_host);
+  // We can't do anything sensible with no hosts.
+  if (weighted_sum == 0) {
+    return;
+  }
+
+  // The minimum possible replication factor is the minimum ring size divided by the weighted sum,
+  // rounded up to make sure we don't have a ring that's too small.
+  const uint64_t min_replication_factor =
+      (min_ring_size / weighted_sum) + (min_ring_size % weighted_sum != 0);
+
+  // The maximum possible replication factor is the maximum ring size divided by the weighted sum,
+  // rounded down to make sure we don't have a ring that's too large. We do have to handle an edge
+  // case here, however, where the ring size bounds are too tight. For example:
+  //   - min_ring_size = 1024
+  //   - max_ring_size = 1024
+  //   - weighted_sum = 5.
+  // In this case, we end up with:
+  //   - min_replication_factor = ceil(1024 / 5) = 205
+  //   - max_replication_factor = floor(1024 / 5) = 204
+  // We played with the idea of handling fractional replication factors, but decided we prefer to
+  // keep the algorithm simpler and more predictable by simply exceeding the max_ring_size in this
+  // case. The "overconstrained ring size bounds" test case reflects this.
+  const uint64_t max_replication_factor =
+      std::max(max_ring_size / weighted_sum, min_replication_factor);
+
+  // If the configuration includes a replication factor, try to use it (or whatever value is closest
+  // to it within bounds). Otherwise, fall back to the old behavior for backward compatibility: size
+  // the ring as small as possible.
+  const uint64_t replication_factor =
+      configured_replication_factor
+          ? std::min(std::max(configured_replication_factor.value(), min_replication_factor),
+                     max_replication_factor)
+          : min_replication_factor;
```

Comment:
Now that we have integer weights, I assume `ASSERT(replication_factor >= 1);` is true. This implies the memory use is going to be something like `O(N_hosts * max(host weight))`. Today, `max(host_weight)` can be something like 2^14 (2^7 * 2^7), so that's 16KB * number of hosts. That seems like it could grow quickly if I have the math correct.

I don't mind the idea of integer replication factors, but I think we need to then scale the weights appropriately if we do this. If it's simpler to do the fractional approach you had before, maybe that is the simplest we can do.

Reply:
Yeah, the way we do it right now, where `effective_weight = host_weight * locality_weight`, gives you a range of 2^7 * 2^7 entries per host. Each entry in the ring consumes more than a byte of memory: it's 8 bytes for the hash, and however big a shared_ptr is, I'm guessing 12 or 16 bytes. So maybe 384KB * number of hosts in the worst case. This is probably not a situation users would ever knowingly subject themselves to, but it could definitely happen.

It's probably a super cool time to mention that we've talked about changing how `effective_weight` is computed to be more like the other load balancers... If you have two localities, with weights of _X_ and _Y_, then you want the first locality to get _X/(X+Y)_ of the traffic and the second locality to get _Y/(X+Y)_ of the traffic, regardless how many hosts are in each. If we tried really hard to stick with integer weights, then the math gets really gross.

So yeah, I felt the fractional approach was the simplest way to cover all possible cases uniformly. I'd be perfectly happy to revert to that, and think of other ways to attempt to simplify... But I sort of want to agree on an approach first so this can all converge. @mattklein123, @htuch - what do you guys think?

Reply:
Yeah, I think when we fix the effective weight calculation to take into account the priority you want to give to locality pick the range of weights will get even wider, so we have to do fractional or scale the weight down.

Reply:
@mattklein123, @htuch - Okay, my conclusion after all this is that I've overconstrained things by adding both the `maximum_ring_size` and `replication_factor` config params. The former is a meaningful number that all users should understand. The latter is a weird number that sort of changes its meaning in the presence of weighting (e.g. if your hosts all have weight 127, what does that make your replication factor?), and it requires users to have a graduate degree in Ketama Studies. So here's my tentative plan:

* Get rid of the `replication_factor` config, and just pick some reasonable number that yields an appropriately sized ring.
* Instead of exposing a `replication_factor` gauge, expose `min_hashes_per_host` and `max_hashes_per_host`. This shows the full range, accounting for weighting, and users can alert if the minimum dips below some threshold.
* Use the fractional algorithm from before, but remove the special handling of the case where ring size bounds are too tight, and maybe steer clear of floating-point (to avoid losing precision).

Let me know if that sounds reasonable.

Reply:
@mergeconflict yes that sounds perfect, I think good compromise between constraining the problem space and user experience. +1

Reply:
Agree.

Reply:
Ok, updated:
* We've got the correct calculation of effective weights now (described in #5982).
* The `replication_factor` config, and all related documentation, is gone.
* We compute the actual ring size basically the same way as in the original implementation (e.g. if you have 8 hosts and a minimum ring size of 10, we'll round up and do 2 hashes per host to get a ring size of 16), and that number is capped by the maximum ring size.
* Tests confirm that no requests are routed to different hosts than before, for existing users.
* The load balancer benchmark shows no regression, most importantly in the `HostLoss` metrics.